service/graph.go

```diff
@@ -31,21 +32,236 @@ var _ pipelines = (*pipelinesGraph)(nil)
+	}
+	return pipelines, nil
+}
+
+// Creates a node for each instance of a component and adds it to the graph
+func (g *pipelinesGraph) createNodes(set pipelinesSettings) error {
+
+	// map[connectorID]pipelineIDs
+	// Keep track of connectors and where they are used.
+	connectorsAsExporter := make(map[component.ID][]component.ID)
+	connectorsAsReceiver := make(map[component.ID][]component.ID)
+
+	for pipelineID, pipelineCfg := range set.PipelineConfigs {
+		for _, recvID := range pipelineCfg.Receivers {
+			if set.Connectors.IsConfigured(recvID) {
+				connectorsAsReceiver[recvID] = append(connectorsAsReceiver[recvID], pipelineID)
+				continue
+			}
+			g.addReceiver(pipelineID, recvID)
+		}
+		for _, procID := range pipelineCfg.Processors {
+			g.addProcessor(pipelineID, procID)
+		}
+		for _, exprID := range pipelineCfg.Exporters {
+			if set.Connectors.IsConfigured(exprID) {
+				connectorsAsExporter[exprID] = append(connectorsAsExporter[exprID], pipelineID)
+				continue
+			}
+			g.addExporter(pipelineID, exprID)
+		}
+	}
+
+	if len(connectorsAsExporter) != len(connectorsAsReceiver) {
+		return fmt.Errorf("each connector must be used as both receiver and exporter")
+	}
+	for connID, exprPipelineIDs := range connectorsAsExporter {
+		rcvrPipelineIDs, ok := connectorsAsReceiver[connID]
+		if !ok {
+			return fmt.Errorf("connector %q must be used as receiver, only found as exporter", connID)
+		}
+		for _, eID := range exprPipelineIDs {
+			for _, rID := range rcvrPipelineIDs {
+				g.addConnector(eID, rID, connID)
+			}
+		}
+	}
+	return nil
+}
+
+func (g *pipelinesGraph) addReceiver(pipelineID, recvID component.ID) {
+	node := newReceiverNode(pipelineID, recvID)
+	if rcvrNode := g.componentGraph.Node(node.ID()); rcvrNode != nil {
+		g.pipelines[pipelineID].addReceiver(rcvrNode)
+		return
+	}
+	g.pipelines[pipelineID].addReceiver(node)
+	g.componentGraph.AddNode(node)
+}
+
+func (g *pipelinesGraph) addProcessor(pipelineID, procID component.ID) {
+	node := newProcessorNode(pipelineID, procID)
+	g.pipelines[pipelineID].addProcessor(node)
+	g.componentGraph.AddNode(node)
+}
+
+func (g *pipelinesGraph) addExporter(pipelineID, exprID component.ID) {
+	node := newExporterNode(pipelineID, exprID)
+	if expNode := g.componentGraph.Node(node.ID()); expNode != nil {
+		g.pipelines[pipelineID].addExporter(expNode)
+		return
+	}
+	g.pipelines[pipelineID].addExporter(node)
+	g.componentGraph.AddNode(node)
+}
+
+func (g *pipelinesGraph) addConnector(exprPipelineID, rcvrPipelineID, connID component.ID) {
+	node := newConnectorNode(exprPipelineID.Type(), rcvrPipelineID.Type(), connID)
+	if connNode := g.componentGraph.Node(node.ID()); connNode != nil {
+		g.pipelines[exprPipelineID].addExporter(connNode)
+		g.pipelines[rcvrPipelineID].addReceiver(connNode)
+		return
+	}
+	g.pipelines[exprPipelineID].addExporter(node)
+	g.pipelines[rcvrPipelineID].addReceiver(node)
+	g.componentGraph.AddNode(node)
+}
+
+func (g *pipelinesGraph) createEdges() {
```

Comment:
I feel that this belongs to `pipelineNodes`. I feel that you are trying to create a global "graph", but I think it is simpler to read/implement if we create a graph where the pipeline is the Node and the Edges are the "Connectors".

So essentially we can probably keep almost all the previous logic to build a pipeline (builtPipeline), only add support for connectors to be receiver/exporters, then add them to a graph and set the edges where we have connectors to determine the initialization order.

Reply:
I've worked out the approximate logic. Forgive the mixed psuedocode format, but I think it would look like this:

### Start (reverse for Shutdown)
- Iterate over global list of exporters, start each one
- Sort the graph topologically. For each node (pipeline) in appropriate order
  - Start each adjacent edge (connector)
  - Start each processor in appropriate order
- Iterate over global list of receivers, start each one

### Build

Some things we use along the way:

```go
// Global indices for quick lookup of pipelines where each connector is used as receiver and exporter
var connectorsAsExporters, connectorsAsReceivers map[component.ID][]component.ID

// Global indices for built receivers, exporters, and connectors, to ensure correct instancing
allReceivers map[component.DataType]map[component.ID]component.Component
allExporters map[component.DataType]map[component.ID]component.Component
allConnectors  map[component.DataType]map[component.DataType]map[component.ID]component.Component

// Basic graph structs & utilities
type Edge struct {
	graph.Edge
	component.ID // connector ID
}
type Node struct {
	component.ID // pipeline ID
}
func(component.ID) int64

// Helper to get first processor or fanout to exporters
func (p *builtPipeline) getFirstConsumer()
```

Before we can build anything, we need to populate the graph so we can reason about topological order. Roughly:
```go
// Cross reference connector indices. Each match represents an edge
for connID, exprPipelineIDs := range connectorsAsExporter {
	for _, eID := range exprPipelineIDs {
		for _, rID := range connectorsAsReceiver[connID] {
			// build Edge, add to graph. Nodes added automatically
		}
	}
}
```

Then we can sort the graph topologically and iterate through pipelines from end to beginning. 

Building exporters is the same, other than identifying the connectors and building those separately.

Building connectors is where it gets a bit ugly. Each connector requires accessing the graph to find the downstream pipeline(s) in which that connector is used as a receiver. This is not necessarily all the edges going out from a pipeline, as there may be multiple connectors in the pipeline. Once we've got the right set of edges, we need to get the appropriate consumer from each pipeline (first processor or fanout), then wrap those together into a fanout. Once built, we can iterate through the edges again and add the connector to all associated `builtPipeline`s, each in the appropriate bucket according to either side of the edge.

```
for each comp in pipelineCfg.exporters
  if comp is a connector
    if conn := allConnectors(comp); conn != nil
      continue
    else
      firstConsumers := []consumers // first consumer of each downstream pipeline
      edgesFromPipeline := graph.From(connector.ID)
        for each edge in edgesFromPipeline
          if edge.ID == connector.ID
            downstreamPipeline := // TODO get pipeline on other side of edge
            firstConsumers.append(downstreamPipeline.getFirstConsumer())
      conn := buildConn(..., fanout(firstConsumers))
      allConnectors.add(conn)
      // iterate through edges again and add to once to each node's builtPipeline.[exporter|receiver]
  else if exp := allExporters(comp); exp != nil
    thisPipeline.exporters.add(exp)
  else
    exp := builtExporter
    allExporters.add(exp)
    thisPipeline.exporters.add(exp)
```

Processors are built the same.
- Build last processor with fanout to `builtPipeline.exporters`
- Build remaining processors in appropriate order


Receivers also need to be separated into receivers and connectors, but otherwise building them is the same. 

When discovering a connector here, we can ignore it because it will be built by another pipeline iteration and added appropriately. (This can probably be inverted so it is the exporters that are deferred.)
```
for each comp in pipelineCfg.receivers
  if comp is a connector
    continue
  else if rcvr := allReceivers(comp); rcvr != nil
    thisPipeline.receivers.add(rcvr)
  else
    rcvr := buildReceiver(...)
    allReceivers.add(rcvr)
    thisPipeline.receivers.add(rcvr)
```

I think this is doable, and probably can be further streamlined. I agree it would be able to use a lot more of the existing code. That said, I think it may not be as straightforward as it first appears.

Personally, I find the global graph easy to reason about because it is built around the actual ontology of the collector - components, pipelines, and a global representation of actual data flow. This other way requires management of artificial indices and lurching between abstraction layers. For example, start and shutdown require working with a global set of one component type, then working within pipelines, then components globally again. And of course building connectors appears to be quite convoluted this way as well.

In any case, I'm far enough into this that I think I can probably get a working implementation together early next week if that's how you'd like to proceed.

Reply:
I started working through the above implementation but it appears that it would be a massive overhaul that is only worse than expected.

What I've proposed in this PR is a clean and working solution that is as well tested as the production code. I've openly pursued this design for three months now and have iterated on it relentlessly to simplify it down to what it is today. I think the suggested overhaul would result in a worse implementation with no meaningful difference in behavior and that it would add unnecessary delay to a feature set that is ready for use by those who opt in. I feel that I have made reasonable efforts to share aspects of the previous implementation but we should expect a temporary state with parallel implementations that diverge in sensible ways.

If there is no other feedback on this PR, I propose that we move forward with my implementation, with the understanding that it may be necessary to work through stylistic changes prior to removal of the feature gate / reduction to one implementation.